[MIOPEN] Add hip graph option to MIOpenDriver

vedenev-amd opened the pull request:
#### Details:


## Motivation

To test kernels work inside the HIP graph.

## Technical Details

[hip graphs explained](https://rocm.docs.amd.com/projects/HIP/en/latest/how-to/hip_runtime_api/hipgraph.html). 
The HIP graph option is implemented only for:  
convolution driver
batch norm driver
activation driver  
- Fixed  shared/ctest/parse_test_categories.py (but it was also fixed later recently in develop)
it is about parsing yaml. Now it can parse complex yaml with nested lists. 
- Improve projects/miopen/src/process.cpp‎, Linux branch
stderr redirected to stdout, so I can use it now in my test to see existence of the HIP Graph.
- Improve projects/miopen/src/process.cpp‎, Windows branch
add all functionality to the Windows branch: environment variables pass, capture output etc
the same like in Linux branch. The problem here that all tests that use miopendriver.exe binary
here via the process.cpp are disabled. There are about 5 tests disabled. So nobody tested this before. 

## Test Plan

Done: [hip_graph_exist.cpp](https://github.com/ROCm/rocm-libraries/blob/8eb1c8b2ecd00b0986e6c70c5e1af16343bf4587/projects/miopen/test/gtest/hip_graph_exist.cpp). 

## Test Result

| # | Test | Time | CI/CD (standard) | Local Run |
|---|------|------|------------------|-----------|
| 0 | conv_hip_graph | 94.83s | Excluded | PASSED |
| 1 | activ_hip_graph | 4.35s | Excluded | PASSED |
| 2 | bnorm_hip_graph | 3.88s | PASSED | PASSED |
| 3 | no_graph | 10.42s | Excluded | PASSED |

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

vedenev committed:
add hip graph option

Commit:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd commented:
The current state of the work:
The miopenConvolutionForward function call is wrapped with the HIP Graph capture.
See the the conv_driver.hpp
There is an issue in the line 2163 in the conv_driver.hpp:
https://github.com/ROCm/rocm-libraries/pull/2865/files#diff-05114d705ab0ae159e4e09ffc6fceb7277d1bcacdbe5b17811aa84c0189bd9bcR2163
`rc = miopenConvolutionForward(GetHandle(),`
the returned rc is 0 zero, so it is not success.
rc=7 (invalid handle)

vedenev-amd commented:
If you use
--time 0
flag then r=0 (all ok, no issues)

vedenev committed:
fix the rc=7 problem, hip Profiling now outside the graph capture region

Commit:
HIP graph helper functions done

Commit:
remove unnecessary tab

vedenev committed:
finished for bn, add test

Commit:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev committed:
fix format

vedenev committed:
fix the naming in the test

vedenev committed:
clang format fix try 2

vedenev committed:
clang format fix try 3

vedenev committed:
clang format fix try 4

vedenev committed:
fix test naming try 2

Commit:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev committed:
clang fromat fix try 5

vedenev committed:
Merge branch 'develop'

Commit:
Trigger CI/CD pipeline

vedenev committed:
Trigger CI/CD pipeline

vedenev committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev committed:
temporary disable some parallelism in CI CD

ScottTodd commented on .github/workflows/therock-test-component.yml:
```diff
@@ -23,6 +23,9 @@ jobs:
     runs-on: ${{ inputs.test_runs_on }}
+    concurrency:
+      group: gpu-test-${{ inputs.platform }}-${{ inputs.amdgpu_families }}
+      cancel-in-progress: false
```
What was the purpose of this [temporary disable some parallelism in CI CD](https://github.com/ROCm/rocm-libraries/pull/2865/commits/e9d6722bed165a1eb95aae42860fa747c2029025) commit?

If this is for testing only, please mark this PR as a draft.

vedenev-amd replied:
Looks like CI CD fails. I want to test it.

vedenev committed:
return back from CI CD

vedenev committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

Commit:
trigger CI CD

vedenev committed:
trigger CI CD 3

Commit:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

JonathanLichtnerAMD commented on projects/miopen/driver/bn_driver.hpp:
```diff
@@ -1139,20 +1140,18 @@ int BatchNormDriver<TInput, Tref, TAcc, TScaleBias, TOut>::RunForwardGPU()
         // if run fwd train
         if(forw == 1)
         { // training only
-            eAF = static_cast<Tref>(1.0) / (static_cast<Tref>(i) + static_cast<Tref>(1.0));
```
Was this supposed to change?

amd-bartgips replied:
I think this is related to a similar problem Maksim encountered with "hipGraphing" the Adam optmizer.
That is, `static_cast<Tref>(i)` is not allowed to be used within the graph, as `i` changes?
It looks like this change affects the bn_driver in general, which I am not sure is desired.

If it is true that without this change hipgraph cannot work, then maybe we should hide this behind a check? I.e. only use `eAF = static_cast<Tref>(0.1)` when using the hip graph functionality?

JonathanLichtnerAMD replied:
Yes, I like the idea of using the old method (CMA) when not using hip graph, and then using 0.1 with hip graph.  @vedenev-amd can we pass i to the lambda, but not use it--i.e., use 0.1--if we're doing hip graph capture?  Then we can pass the iteration to the Execute method, and pass that to the lambda function in the non-hip graph case.

JonathanLichtnerAMD replied:
Thanks, the updated change looks good to me

JonathanLichtnerAMD commented on projects/miopen/driver/conv_driver.hpp:
```diff
@@ -2076,8 +2079,7 @@ int ConvDriver<Tgpu, Tref>::RunForwardGpuFind(const bool is_transform)
-    for(int i = 0; i < num_iterations; i++)
-    {
+    int return_code = HipGraphCapture([&]() -> int {
         if(init_output_nan)
```
I think this if statement (to init with NaNs) should move inside the for loop on line 2106, otherwise the output tensor won't be initialized to NaNs for each iteration.

JonathanLichtnerAMD commented on projects/miopen/driver/driver.cpp:
```diff
@@ -54,3 +70,120 @@ GPUMem::Check Driver::GetGpuBufferCheck(const InputFlags& inflags) const
+    }
+
+    he = hipGraphInstantiate(&hipGraphExec, hipGraph, nullptr, nullptr, 0);
+    if(he != hipSuccess)
```
In the hip graph capture documentation, I believe you can destroy the hip graph unconditionally at this point:

```
    // The graph template can be deleted after the instantiation if it's not needed for later use
    HIP_CHECK(hipGraphDestroy(graph));
```

This will save the later calls to destroy this...

JonathanLichtnerAMD commented on projects/miopen/driver/driver.cpp:
```diff
@@ -54,3 +70,120 @@ GPUMem::Check Driver::GetGpuBufferCheck(const InputFlags& inflags) const
+    // if(rc != miopenStatusSuccess)
+    // {
+    //     return rc;
+    // }
```
Is the commented out code needed for any reason?  If not, please remove it...

JonathanLichtnerAMD commented on projects/miopen/driver/driver.cpp:
```diff
@@ -54,3 +70,120 @@ GPUMem::Check Driver::GetGpuBufferCheck(const InputFlags& inflags) const
+    // int rc = functPtr();
+    // if(rc != miopenStatusSuccess)
+    // {
+    //     return rc;
+    // }
+
+    hipError_t he = hipStreamBeginCapture(q, hipStreamCaptureModeGlobal);
+    if(he != hipSuccess)
+    {
+        return miopenStatusInternalError;
+    }
+
+    int rc = functPtr();
+
+    if(rc != miopenStatusSuccess)
+    {
+        hipStreamEndCapture(q, &hipGraph);
+        return rc;
+    }
+
+    he = hipStreamEndCapture(q, &hipGraph);
+    if(he != hipSuccess)
+    {
+        return miopenStatusInternalError;
+    }
```
This code can be simplified to the following:
```
    int rc = functPtr();
    he = hipStreamEndCapture(q, &hipGraph);
    if(rc != miopenStatusSuccess || he != hipSuccess) {
        hipGraphDestroy(hipGraph);
        if(rc != miopenStatusSuccess)
            return rc;
        return miopenStatusInternalError;
    }
```
since we unconditionally call hipStreamEndCapture, and we don't want to forget to destroy the hip graph.

JonathanLichtnerAMD commented on projects/miopen/driver/driver.cpp:
```diff
@@ -54,3 +70,120 @@ GPUMem::Check Driver::GetGpuBufferCheck(const InputFlags& inflags) const
+    if(use_hip_graph)
+    {
+        hipGraphIsProfilingToRestore = miopen::deref(GetHandle()).IsProfilingEnabled();
+        miopenEnableProfiling(GetHandle(), false);
```
I think you can just override the profiling and don't need to restore this value

JonathanLichtnerAMD commented on projects/miopen/driver/conv_driver.hpp:
```diff
@@ -2076,8 +2079,7 @@ int ConvDriver<Tgpu, Tref>::RunForwardGpuFind(const bool is_transform)
 
-    for(int i = 0; i < num_iterations; i++)
-    {
+    int return_code = HipGraphCapture([&]() -> int {
```
The naming of HipGraphCapture is a little misleading, since this only does a capture in when hip graphs are enabled via use_hip_graph.  Could we rename these to be more generic, e.g., something like HipGraphCapture -> CaptureKernel, HipGraphExecute -> ExecuteKernel, etc?

JonathanLichtnerAMD commented on projects/miopen/driver/adam_driver.hpp:
```diff
@@ -314,6 +315,8 @@ int AdamDriver<Tgpu, Tref, Tgrad>::AddCmdLineArgs()
     inflags.AddInputFlag(
         "wall", 'w', "0", "Wall-clock Time Each Layer, Requires time == 1 (Default=0)", "int");
 
+    AddHipGraphFlag(inflags);
```
Let's remove the hip graph option from adam_driver.hpp since this doesn't do anything

JonathanLichtnerAMD commented on projects/miopen/test/gtest/hip_graph_exist.cpp:
```diff
@@ -0,0 +1,394 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
```
Let's use the new copyright:
```
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
```

JonathanLichtnerAMD commented on projects/miopen/test/gtest/hip_graph_exist.cpp:
```diff
@@ -0,0 +1,394 @@
+            }
+        }
+    }
+};
```
Thank you for writing this test.  It's really nice that this functionality is tested.

One aspect we could potentially improve on is that the unit test is a little noisy (to the output).  Could we use some of the gtest mechanisms to capture stderr or stdout?  E.g.,

```
testing::internal::CaptureStderr();

...

auto output = testing::internal::GetCapturedStderr();
```

Then statements like
```
EXPECT_THAT(output, Not(testing::HasSubstr("Warning [IsEnoughWorkspace]")));
```

JonathanLichtnerAMD commented on projects/miopen/test/gtest/hip_graph_exist.cpp:
```diff
@@ -0,0 +1,394 @@
+             "bnorm_hip_graph",
+             true},
+            {"conv",
+             "-n 4 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 "
```
I timed these tests and they took about 10 seconds to run.  Can we smaller shapes and maybe `--iter=1` (so we only do one iteration) and see if that speeds up the tests?

JonathanLichtnerAMD commented on projects/miopen/test/gtest/hip_graph_exist.cpp:
```diff
@@ -0,0 +1,394 @@
+        temp_dir = std::string(temp_path) + "miopen_hip_graph_test";
+#else
+        temp_dir    = "/tmp/miopen_hip_graph_test";
+#endif
```
Can we use std::filesystem::temp_directory_path to get the /tmp directory path?  I think that should work for both Windows and Linux

vedenev committed:
eAF fix, biases wrapping

vedenev-amd committed:
Test commit for author verification

vedenev-amd committed:
Author change test

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
test fix and others

vedenev-amd committed:
Apply clang-format 18.1.4 to fix formatting issues

vedenev-amd committed:
keep adam driver unchencged

vedenev-amd committed:
clang fromat 12 for miopen

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
removed restoring profiling state, renaming

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
return CMA for non Hip Graph, keep 0.1 for HIP graph, in bn_norm

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
fix after mege bugs

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
empty commit to trigger CI CD

vedenev-amd committed:
empty commit to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
empty commit to trigger CI CD

vedenev-amd committed:
empty commit to trigger CI CD

vedenev-amd committed:
empty commit to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

JonathanLichtnerAMD committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
return Q for the --ping_pong_buffers, J for --use_hip_graph

JonathanLichtnerAMD committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
fix potential bug with CaptureStderr

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
empty commit to trigger CI CD

vedenev-amd committed:
use standart testing tools

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
format

vedenev-amd committed:
fix for nodiscard new CICD's staff

vedenev-amd committed:
format

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
fix test naming

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
merge develop to trigger CI CD

vedenev-amd committed:
test runs loccally, exclude 3 test from Ci CD, only one fastest in the CI CD, potential corrupted memory bug fixed

vedenev-amd committed:
rocprofv3 use, try 2 to exclude the 3 tests

vedenev-amd committed:
undo changes in the test_categories.yaml, try to run all 4 tet in the CI CD

vedenev-amd committed:
Merge develop into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
not use rocprof, use drvier, exists?

vedenev-amd committed:
format

vedenev-amd committed:
Merge remote-tracking branch 'origin/develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
fix popen pclose for windows, merge develop

vedenev-amd committed:
try 3 to make 3/4 exclusion for the tests: add correct GPU_... rule to the yaml, and exclusion of 3 tests, the same style like in GPU_Softmax

vedenev-amd committed:
fix the parse_test_categories.py (nested lists bug) to run the hip graph test in Ci CD

vedenev-amd committed:
remove filtering with the GPUs maks for hip graph test

vedenev-amd committed:
fix bug in windows testing, improve windows testing, remove  LD_LIBRARY_PATH

vedenev-amd committed:
format

vedenev-amd committed:
Merge develop branch to trigger CI/CD

vedenev-amd committed:
a fix: now stderr rederected to the stdout in the Process, enable no_graph test to se if it hangs in the Ci CD

vedenev-amd committed:
temporrary enable gemm test for windows in the CI CD to see if it works

vedenev-amd committed:
use _popen in windows, will it fix hang?

vedenev-amd committed:
disable back the gem test, dispate the it not included in previuse commit by some unclonw reason, live only /2 hip graph test for speed, now all fixed for the windows?

vedenev-amd committed:
Merge remote-tracking branch 'origin/develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
fix linux test exclude

vedenev-amd committed:
[MIOpen] Fix Windows testing infrastructure for driver tests

- Fix process.cpp: Use popen/pclose instead of CreateProcess for Windows
  - Enables output capture with stderr redirected to stdout
  - Enables environment variable overrides
  - Simplifies Windows process handling

- Fix miopendriver_common.hpp: Proper MIOpenDriver.exe path detection on Windows
  - Use GetModuleFileNameW to get executable path
  - Add .exe extension for Windows

vedenev-amd committed:
fixes for Randy's comments

vedenev-amd committed:
Merge branch 'users/mvedenev-amd/improve_CI_CD' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge develop to fix WinoFury test failures

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge remote-tracking branch 'origin/develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

vedenev-amd committed:
disable Dockerfile change

vedenev-amd committed:
test exclussion on the cpp level to be able to exclude in jenkins also

vedenev-amd committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph

JonathanLichtnerAMD reviewed (approved):
LGTM.  Please update the PR description before merging, i.e., just summarize the changes, i.e., remove the 

-   How much time is needed:
- What is need to do:

sections, etc

JonathanLichtnerAMD committed:
Merge branch 'develop' into silo/miopen/feature/add-hip-graph
